### PR TITLE
feat(index.jsx): Updating to support react-cosmos3

### PR DESCRIPTION
**Pull request by scottdickerson**
test(proxy.spec & default.spec) Adding testcases to reach full code
coverage through unit tests (and pass build)
feat(package.json): updating peer deps and dev deps to react-cosmos3,
added enzyme to support new unit test

### CODE REVIEW

**Comment by NiGhTTraX** [comment]
@scottdickerson I think the unit test that checks the default values is unnecessary as we should be getting coverage from the acceptance tests. It's also better to not have the unit tests check more than 1 component at a time - in this case we're checking both the proxy and the `Docs` component.

Also, there's an [error](https://travis-ci.org/NiGhTTraX/react-docs-cosmos-proxy/builds/353953221#L3885) on Travis about compiling the Cosmos config. Not sure if anything's changed with Cosmos@3 related to this...

P.S.: Please don't commit the package lock for now, I think it should be added when the project is in a releasable state.

**Comment by scottdickerson** [comment]
@NiGhTTraX I made all the changes you requested, but am now getting null in my nyc coverage json files and nyc is throwing an error trying to rollup the coverage reports.  Any ideas?
`emoving acceptance_playground_1         ... done
Removing network acceptance_default
/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/yargs.js:1100
      else throw err
           ^

TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at SourceMaps.reloadCachedSourceMaps (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/lib/source-maps.js:37:10)
    at /Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:528:22
    at Array.map (<anonymous>)
    at NYC.loadReports (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:517:16)
    at NYC._getCoverageMapFromAllCoverageFiles (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:424:8)
    at NYC.report (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/index.js:442:18)
    at Object.exports.handler (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/lib/commands/report.js:38:7)
    at Object.runCommand (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/lib/command.js:228:22)
    at Object.parseArgs [as _parseArgs] (/Users/scottsd/Documents/react-docs-cosmos-proxy/node_modules/nyc/node_modules/yargs/yargs.js:1013:30)
npm ERR! Test failed.  See above for more details.`

**Comment by scottdickerson** [comment]
@NiGhTTraX This is ready to go but FYI our acceptance tests are not generating code coverage reports (I added a warning to catch and report this case).  I went ahead and added the necessary unit tests to pass the code coverage threshold.

**Comment by scottdickerson** [comment]
@NiGhTTraX ok it's ready for you to review...

File: package.json [hunk 1]
```diff
@@ -28,8 +28,8 @@
     "_test:integrated:debug": "karma start tests/integrated/karma.debug.config.js",
     "test:acceptance": "./tests/acceptance/run.sh",
     "test:acceptance:debug": "./tests/acceptance/debug.sh",
-    "_test:acceptance": "mocha --opts tests/acceptance/mocha.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
-    "_test:acceptance:debug": "mocha --opts tests/acceptance/mocha.debug.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
+    "_test:acceptance": "mocha --require ignore-styles --opts tests/acceptance/mocha.opts tests/acceptance/setup.js --recursive tests/acceptance/specs",
```
Comment (NiGhTTraX) [inline package.json:31]: Please move the `--require` param to the opts file - this way we don't invalidate the Docker cache.

File: playground/components/acceptance/.babelrc [hunk 2]
```diff
@@ -2,6 +2,7 @@
   "extends": "../../../.babelrc",
 
   "plugins": [
-    "react-docgen"
+    "react-docgen",
+    "istanbul"
```
Comment (NiGhTTraX) [inline playground/components/acceptance/.babelrc]: The "acceptance" components for the playground don't need to be instrumented, they're basically fixtures. If the coverage reports are failing because of this that's fine, we'll fix it later.

File: playground/cosmos.config.js [hunk 3]
```diff
@@ -1,16 +1,8 @@
...
-    path.join(__dirname, '../src/components'),
-    path.join(__dirname, 'components')
-  ],
-
-  fixturePaths: [
-    path.join(__dirname, 'fixtures')
-  ],
-
+module.exports = {
+  fileMatch: '**/{__fixture?(s)__,fixture?(s)}/**/*.{js,jsx,ts,tsx}',
```
Comment (NiGhTTraX) [inline playground/cosmos.config.js:5]: Please remove the TypeScript extensions :)

File: src/index.jsx [hunk 4]
```diff
@@ -60,3 +60,5 @@ export default ({
     </div>;
   }
 };
+
+export default createDocsProxy;
```
Comment (NiGhTTraX) [inline src/index.jsx:64]: There's no need for this to be separated from the assignment since we assign static members using class fields. Please move it back up.

File: tests/acceptance/helpers.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ export async function loadFixture(component, fixture) {
 
   // Switch focus to the only frame on the page, the one displaying the
   // component.
+  await browser.waitForVisible('iframe', 15 * 1000);
```
Comment (NiGhTTraX) [inline tests/acceptance/helpers.js:15]: Why is this necessary? The tests setup already [waits](https://github.com/NiGhTTraX/react-docs-cosmos-proxy/blob/16e64d54dbb4c436233623363605156b0f777987/tests/acceptance/setup.js#L54) for the playground to load. Same question for the `checkForVisualChanges` change below.

File: tests/acceptance/setup.js [hunk 6]
```diff
@@ -48,23 +48,25 @@ before('Connecting to Selenium', function() {
   return client;
 });
 
-beforeEach('Waiting for app to render', function() {
+beforeEach('Waiting for app to render', async function() {
```
Comment (NiGhTTraX) [inline tests/acceptance/setup.js:51]: There's no need to declare this as `async` because we're not `await`-ing anything, it's just returning a promise.

File: tests/acceptance/setup.js [hunk 7]
```diff
@@ -48,23 +48,25 @@ before('Connecting to Selenium', function() {
...
   });
 
   const name = this.currentTest.fullTitle().replace(/\//g, '_');
-
   fs.writeFileSync(
     path.join(__dirname, 'results', 'coverage', `${BROWSER}_${name}.json`),
-    coverage
+    coverage || ''
   );
+  if (!coverage) {
```
Comment (NiGhTTraX) [inline tests/acceptance/setup.js:67]: When would this happen? I would rather understand the problem than add warnings.

File: tests/acceptance/specs/fixtures.spec.js [hunk 8]
```diff
@@ -1,23 +1,46 @@
...
-      Object.keys(fixtures[component])
-        .forEach(fixture => {
-          it(fixture, async () => {
-            await loadFixture(component, fixture);
-          });
+function getFixtureModules(fixtureFiles) {
+  return fixtureFiles.reduce(
+    (acc, f) => ({
+      ...acc,
+      [f.filePath]: importModule(require(f.filePath)) // eslint-disable-line
```
Comment (NiGhTTraX) [inline tests/acceptance/specs/fixtures.spec.js:20]: Please disable only the necessary rule.

File: tests/acceptance/specs/fixtures.spec.js [hunk 9]
```diff
@@ -1,23 +1,46 @@
...
+    }
+    ),
+    {}
```
Comment (NiGhTTraX) [inline tests/acceptance/specs/fixtures.spec.js:23]: These 3 lines can be joined :)

File: tests/acceptance/specs/fixtures.spec.js [hunk 10]
```diff
@@ -1,23 +1,46 @@
...
+  );
+}
+
+describe('fixtures test', async() => {
```
Comment (NiGhTTraX) [inline tests/acceptance/specs/fixtures.spec.js:27]: I'm not sure if Mocha waits for this. As I recall defining tests needs to happen sync. Moreover, if it works, this is changing the previous behavior by introducing a prefix in the test names, thus forcing a rename of the screenshots.

Truth be told, I have the visual tests working with Cosmos@3 stashed somewhere and I just need to find the time to commit it. Since that's WIP and I'm fine with having failing acceptance tests in master I would rather not continue with making them work.

File: tests/unit/components/docs-table/columns/default.spec.js [hunk 11]
```diff
@@ -8,4 +8,12 @@ describe('Default column', () => {
     const $default = $render(<Default docs={{ required: false }} />);
     expect($default.text()).to.be.empty;
   });
+  it('show required if set', () => {
+    const $default = $render(<Default docs={{ required: true }} />);
+    expect($default.text()).to.equal('Required');
+  });
+  it('show default value if set', () => {
```
Comment (NiGhTTraX) [inline tests/unit/components/docs-table/columns/default.spec.js:15]: Please separate `it` blocks with a new line :)

File: tests/unit/components/proxy.spec.js [hunk 12]
```diff
@@ -42,6 +46,15 @@ describe('DocsProxy', () => {
       render(<DocsProxy {...props} />);
     });
 
+    it('if no docs component passed, should render default', () => {
+      const FakeDocsTable = createSpy({ name: 'FakeDocsTable' });
+      createDocsProxy.__Rewire__('DocsTable', FakeDocsTable);
```
Comment (NiGhTTraX) [inline tests/unit/components/proxy.spec.js:51]: `DocsTable` being used by the proxy as a default value is an implementation detail - hence why something like rewire or enzyme is needed in the tests - so I don't want to unit test that. The acceptance tests will cover it by checking the whole thing end to end.

**Review by NiGhTTraX — COMMENTED** [review]
@scottdickerson thanks for all the hard work. Sorry to nitpick about the tests. In fact, I'm just gonna merge like this and worry about them another day. I need to bring in the latest changes from [react-test-buffet](https://github.com/NiGhTTraX/react-test-buffet) anyway.

File: tests/acceptance/helpers.js (hunk 5 above)
Comment (scottdickerson) [reply tests/acceptance/helpers.js:15]: I think the iframe was loading a little slower than the main playground header.  (this is my guess anyway, these timing issues were hard to resolve)

File: src/index.jsx (hunk 4 above)
Comment (scottdickerson) [reply src/index.jsx:64]: I was getting an eslint error I think on this because the function wasn't named.
